### PR TITLE
Add RequireRemote#load to load external Ruby scripts from server

### PR DESCRIPTION
**Pull request by ledsun**
## Motivation

To run a ruby.wasm application consisting of multiple Ruby scripts, the developer needs to enumerate multiple Ruby scripts.
Like below:

```html
<script type="text/ruby" src="lib_a"></script>
<script type="text/ruby" src="lib_b"></script>
<script type="text/ruby" src="main"></script>
```

I want to automatically resolve dependencies of Ruby scripts according to the definition of require_relative.

In `main.rb`:

```ruby
require_relative 'lib_b'
...
```

In `lib_a.rb`

```ruby
require_relative 'lib_b'
...
```

## Feature

This pull request provides `RequireRemote#load` method.
`RequireRemote#load` loads Ruby scripts from server.
This is used to emulate `require_relative`.
The simplest expamle is below:

```ruby
require 'js/require_remote'

module Kernel
  def require_relative(path) = JS::RequireRemote.instance.load(path)
end
```

The argument of the `RequireRemote#load` method is treated as a relative URL, and the dependent ruby script is fetched from that URL.
For example, `RequireRemote#load 'lib_a'` will fetch the script from the URL `./lib_a.rb`.

### CODE REVIEW

File: ext/js/lib/js/browser.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
+require "singleton"
+require_relative "./url_resolver"
+
+module JS
+  ScriptLocation = Data.define(:url, :filename)
+
+  class Browser
```
Comment (ledsun) [inline ext/js/lib/js/browser.rb:7]: I feel that the name "browser" represents a role that is a bit too broad. I am trying to think of a better name.

**Review by kateinoigakukun — COMMENTED** [review]
Thank you for working on the feature and sorry for not responding for a while :bow: 
Several major comments:

1. Can we use another name for this feature to avoid patching `require_relative`?  The method is already patched by several places and I don't want to introduce new complexities here. How about `require_remote`?
2. I'd like to make this feature opt-in since `fetch` is a part of Web API and not generally available on non-Web JS runtimes. Could you remove `require_relative "js/loader.rb"` statement from js.rb and let users explicitly require the file by `require "js/loader"`?
3. What happens if "./foo.rb" is redirected to "./bar.rb" at HTTP level and a user tries to load both? This is similar to file system symlink with `require_relative` and it loads only one of them.
4. I think, `loader` is a little bit too general word. To express it loads remote contents, how about `js/remote_loader.rb`, or `js/require_remote.rb`?

**Comment by kateinoigakukun** [comment]
If you face build stuck on your local, please rebase upon the latest main :pray:

**Comment by ledsun** [comment]
Thanks for letting me know. I was stuck in the following part of the build and it has been resolved.

>   ==> make -j1 install_sw DESTDIR\=/home/ledsun/ruby.wasm/build/wasm32-unknown-wasi/openssl-3.0.5/opt

File: ext/js/lib/js/require_remote/response_handler.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      end
+
+      # Execute the body of the response and record the URL.
+      def execute_and_record(response, location)
+        if response[:status].to_i == 200
+          # Check if the redirected URL has already evaluated.
+          return false if url_evaluated?(response[:url].to_s)
+
+          code = response.text().await.to_s
+          evaluate(code, location)
```
Comment (ledsun) [inline ext/js/lib/js/require_remote/response_handler.rb:18]: TODO: The URL to be used as the basis for the relative path is the URL after the redirect.

File: packages/npm-packages/ruby-wasm-wasi/test-e2e/integrations/browser-script.spec.ts [hunk 3]
```diff
@@ -60,4 +60,85 @@ if (!process.env.RUBY_NPM_PACKAGE_ROOT) {
...
+        require 'js/require_remote'
+        JS::RequireRemote.instance.load 'error_on_load_twice'
+        JS.global.checkResolved JS::RequireRemote.instance.load 'error_on_load_twice'
+      </script>
+     `);
+
+      expect(await resolve()).toBe(false);
+    });
+
+    test("JS::RequireRemote#load identifies by URL after redirect", async ({
```
Comment (ledsun) [inline packages/npm-packages/ruby-wasm-wasi/test-e2e/integrations/browser-script.spec.ts:100]: Under investigation

This test fails.
In the integration test, response#url returns the URL before the redirect.
I do not know the cause.

File: packages/npm-packages/ruby-wasm-wasi/test-e2e/support.ts [hunk 4]
```diff
@@ -23,8 +24,33 @@ export const setupProxy = (context: BrowserContext) => {
...
+      "dist",
+      relativePath,
+    );
+
+    if (relativePath.match("error_on_load_twice.rb")) {
+      route.fulfill({
+        body: `raise 'load twice' if defined? ALREADY_LOADED
+               ALREADY_LOADED = true`,
+        contentType: "text/ruby",
+      });
```
Comment (ledsun) [inline packages/npm-packages/ruby-wasm-wasi/test-e2e/support.ts:38]: TODO: I feel uncomfortable writing responses for individual tests in this file that is shared by multiple tests.
I would like to separate them into separate files.

File: packages/npm-packages/ruby-wasm-wasi/test-e2e/integrations/js-require-remote.spec.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import path from "path";
+import { test, expect } from "@playwright/test";
+import { setupDebugLog, setupProxy, resolveBinding } from "../support";
+
+if (!process.env.RUBY_NPM_PACKAGE_ROOT) {
+  test.skip("skip", () => {});
+} else {
+  test.beforeEach(async ({ context }) => {
+    setupDebugLog(context);
+    setupProxy(context, (route, relativePath, mockedPath) => {
```
Comment (kateinoigakukun) [inline packages/npm-packages/ruby-wasm-wasi/test-e2e/integrations/js-require-remote.spec.ts:11]: Sorry for the ambiguous function name but could you keep `setupProxy` only for mocking CDN stuff and just add new Playright routing?
I think this kind of customizing injection leads to a bloat of function responsibilities.

**Review by kateinoigakukun — COMMENTED** [review]
Would you mind adding unit test also to cover more cases? I wonder what should happen for the following case:

```
.
├── lib
│   ├── a.rb -- require_relative "b.rb"
│   └── b.rb
└── main.rb  -- require_relative "lib/a.rb"
```

Is the current implementation able to load `b.rb` well?

File: packages/npm-packages/ruby-wasm-wasi/test-e2e/integrations/js-require-remote.spec.ts (hunk 5 above)
Comment (ledsun) [reply packages/npm-packages/ruby-wasm-wasi/test-e2e/integrations/js-require-remote.spec.ts]: I stopped extending `setupProxy`.
Playwright's routing was sufficient for this test cases.
I was wondering if there were duplicate conditions that could not be handled. Thanks for the advice.

I have left the modification that made mockedPath a local variable. Please let me know if I should delete it. I will fix it.

**Comment by ledsun** [comment]
Added test case to run require_relative recursively. 
Please let me know if there are other test cases that you think we should add.

File: ext/js/lib/js/require_remote/url_resolver.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        if relative_feature.end_with?(".rb")
+          relative_feature
+        else
+          "#{relative_feature}.rb"
+        end
+      end
+
+      # Return a URL object of JavaScript.
+      def resolve(relative_filepath)
+        JS.global[:URL].new relative_filepath, @url_stack.last
```
Comment (kateinoigakukun) [inline ext/js/lib/js/require_remote/url_resolver.rb:41]: I think we need to carefully handle relative path -> absolute path resolution here. IIUC JavaScript's `new URL(path, base)` does not behave as we expect for `Kernel#require_relative` (and also `File.expand_path`).

Here is an example:
| `path, base` | `File.expand_path(path, base)` | `new URL(path, "https://example.com" + base)` |
|:-------------------|:---------------------|:---------------------|
| path="b.rb", base="/lib" | `/lib/b.rb` | `https://example.com/b.rb`
| path="./b.rb", base="/lib" | `/lib/b.rb` | `https://example.com/lib/b.rb`

I think the test suite of `File.expand_path` would be helpful for you to see what should be cared https://github.com/ruby/ruby/blob/master/spec/ruby/core/file/expand_path_spec.rb

File: packages/npm-packages/ruby-wasm-wasi/test-e2e/integrations/js-require-remote.spec.ts (hunk 5 above)
Comment (kateinoigakukun) [reply packages/npm-packages/ruby-wasm-wasi/test-e2e/integrations/js-require-remote.spec.ts]: Sounds reasonable to me, thanks!

File: ext/js/lib/js/require_remote/url_resolver.rb (hunk 6 above)
Comment (ledsun) [reply ext/js/lib/js/require_remote/url_resolver.rb:41]: I believe this behavior of the  JavaScript's URL constructor is as expected.

In most cases, I assume that the first value for `base` is the URL of an HTML file, such as http://exapmle.com/index.html, or a URL such as http://exapmle.com/index without the extension.
Also, once the ruby script is loaded, the `base` value will be the URL of the ruby script, such as http://exapmle.com/a.rb.

So we want the string after the last `/` in the `base` value to be ignored.

Comment (kateinoigakukun) [reply ext/js/lib/js/require_remote/url_resolver.rb:41]: Ah, ok. I misunderstood the second argument. If the second argument is always a URL to a file, then it makes sense to me.

Comment (kateinoigakukun) [reply ext/js/lib/js/require_remote/url_resolver.rb:41]: Anyway, unit tests covering this area would be preferable.

Comment (ledsun) [reply ext/js/lib/js/require_remote/url_resolver.rb:41]: I wrote unit testsfor JS::RequireRemote::URLResolver.
Please let me know if there are any tests that you think we should add.

**Pull request merged**